### PR TITLE
[SPARK-32940][SQL] Collect, first and last should be deterministic aggregate functions

### PR DESCRIPTION
**Pull request by tanelk**
### What changes were proposed in this pull request?

Collect, first and last have mistakenly been marked as non-deterministic. They are actually deterministic iff their child expression is deterministic.

For example collect was marked as non-deterministic in #14749. The reasoning was that its output depends on the actual order of input rows. Although it is correct that these aggregators depend on the order of input rows, it does not make them non-deterministic.

In `EliminateSorts` optimizer rule, there is a method `isOrderIrrelevantAggs`, that lists all aggregators that do not depend on their input row order. Collect, first and last are correctly not listed there. 
An aggregator would be non-deterministic if its output for a group would depend on previous groups it has aggregated - I can't think of any practical examples of this kind of aggregator in Spark.

An analogous aggregator to these would be sum on float and double datatype  - its result does depend on the order of its inputs, but is deterministic. Another similar aggregates are the `max_by` and `min_by` - deterministic functions, that can return different results when the order of rows changes.

### Why are the changes needed?

The optimizer rule `PushPredicateThroughNonJoin` can work in more cases. 

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

UT

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparisonSuite.scala [hunk 1]
```diff
@@ -120,10 +120,10 @@ class UnwrapCastInBinaryComparisonSuite extends PlanTest with ExpressionEvalHelp
...
 
   test("unwrap cast should skip when expression is non-deterministic or foldable") {
-    Seq(positiveInt, negativeInt).foreach (v => {
-      val e = Cast(First(f, ignoreNulls = true), IntegerType) <=> v
+    Seq(positiveLong, negativeLong).foreach (v => {
+      val e = Cast(SparkPartitionID(), LongType) <=> v
       assertEquivalent(e, e, evaluate = false)
-      val e2 = Cast(Literal(30.toShort), IntegerType) >= v
+      val e2 = Cast(Literal(30), LongType) >= v
       assertEquivalent(e2, e2, evaluate = false)
```
Comment (tanelk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparisonSuite.scala:127]: There was no other non-deterministic expression, that can return a short, so I had to change this test a bit.

**Comment by tanelk** [comment]
@cloud-fan, you reviewed the related pull request (although years back).

**Comment by dongjoon-hyun** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/First.scala [hunk 2]
```diff
@@ -63,9 +63,6 @@ case class First(child: Expression, ignoreNulls: Boolean)
 
   override def nullable: Boolean = true
 
-  // First is not a deterministic function.
-  override lazy val deterministic: Boolean = false
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/First.scala:67]: I think you may need to update the `note` above and says like "The function can be non-deterministic because its results depend on the order of input rows which are usually non-deterministic after a shuffle." You might need to update `functions.py`, `functions.R` and `functions.scala`

**Comment by cloud-fan** [comment]
> In EliminateSorts optimizer rule, there is a method isOrderIrrelevantAggs, that lists all aggregators that do not depend on their input row order. Collect, first and last are correctly not listed there.

hmm, it's pretty weird if we list `first/last` there. Removing sort will definitely change the query result, doesn't it?

**Comment by tanelk** [comment]
> > In EliminateSorts optimizer rule, there is a method isOrderIrrelevantAggs, that lists all aggregators that do not depend on their input row order. Collect, first and last are correctly not listed there.
> 
> hmm, it's pretty weird if we list `first/last` there. Removing sort will definitely change the query result, doesn't it?

Sorry if I didn't word it correctly - these are not listed there.  I tried to exemplify the difference between deterministic and order irrelevant.

**Comment by hvanhovell** [comment]
Maybe I am missing something here. AFAIK the problem with First/Last/CollectList methods is that we can't control how results are merged. This depends on how we shuffle fetches results and this is not deterministic.

**Comment by tanelk** [comment]
> Maybe I am missing something here. AFAIK the problem with First/Last/CollectList methods is that we can't control how results are merged. This depends on how we shuffle fetches results and this is not deterministic.

You are 100% correct. As a user, this is how I would also understand the term deterministic.
But, internally deterministic has different meaning - by this definition `Sum` should be also non-deterministic if its input type is float or double. 

I'll copy our internal definition:
```
   * Note that this means that an expression should be considered as non-deterministic if:
   * - it relies on some mutable internal state, or
   * - it relies on some implicit input that is not part of the children expression list.
   * - it has non-deterministic child or children.
   * - it assumes the input satisfies some certain condition via the child operator.
```
For aggregation expressions the internal state part can introduce extra confusion - of course all of them have some internal state about the current group they are aggregating (running count, largest value seen  so far, etc), but they do not "remember" the previous groups they have aggregated. 

There is a separate optimizer rule `EliminateSorts`, that keeps track of aggregators, that do not depend on input order - max, count, etc. But these are a subset of all deterministic aggregators.

For context, why this is relevant:
A snippet from `PushPredicateThroughNonJoin`
https://github.com/apache/spark/blob/c336ddfdb81dd5c27fd109d62138dc129a02c30b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala#L1138-L1141

Basically this case will filter out groups in the aggregation before aggregating the values. Within one group the aggregator will still see all the same rows in the same order, but it would not see the groups, that were filtered out. This would change the output of an aggregator, that remembers previous groups (non-deterministic), but it would not change the output of an aggregator, that only cares about the current group (deterministic, but possibly order relevant).

**Comment by tanelk** [comment]
Pinging @maropu

**Comment by cloud-fan** [comment]
There is inevitable randomness in the input of aggregate functions, because the shuffle reader may produce data with random orders, and we are not able to completely eliminate this randomness. For example, even `sum` has randomness, as adding up floating values with different orders can lead to slightly different results.

I don't think `first/last/collect` has a significant difference and +1 to mark them as deterministic.

**Comment by cloud-fan** [comment]
retest this please

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparisonSuite.scala [hunk 3]
```diff
@@ -183,10 +182,10 @@ class UnwrapCastInBinaryComparisonSuite extends PlanTest with ExpressionEvalHelp
   }
 
  test("unwrap cast should skip when expression is non-deterministic or foldable") {
-    Seq(positiveInt, negativeInt).foreach(v => {
-      val e = Cast(First(f, ignoreNulls = true), IntegerType) <=> v
+   Seq(positiveLong, negativeLong).foreach (v => {
+     val e = Cast(SparkPartitionID(), LongType) <=> v
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparisonSuite.scala:186]: can we use `Rand`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateDistinctSuite.scala [hunk 4]
```diff
@@ -39,11 +39,15 @@ class EliminateDistinctSuite extends PlanTest {
...
+    First(_, ignoreNulls = true),
+    First(_, ignoreNulls = false),
+    Last(_, ignoreNulls = true),
+    Last(_, ignoreNulls = false),
     CollectSet(_: Expression)
   ).foreach {
     aggBuilder =>
       val agg = aggBuilder('a)
-      test(s"Eliminate Distinct in ${agg.prettyName}") {
+      test(s"Eliminate Distinct in ${agg.toString}") {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateDistinctSuite.scala:50]: nit: just $agg

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!
